Implement trusted types enforcement for script elements

lukewarlow opened the pull request:
#### f33204d52243ae13948a2dee417d69e25442c2d8
<pre>
Implement trusted types enforcement for script elements
<a href="https://bugs.webkit.org/show_bug.cgi?id=269365">https://bugs.webkit.org/show_bug.cgi?id=269365</a>

Reviewed by Darin Adler and Ryosuke Niwa.

Add script text string to script element. This is used to ensure that script elements can&apos;t be manipulated in untrusted ways.

See <a href="https://w3c.github.io/trusted-types/dist/spec/#htmlscriptelement-script-text">https://w3c.github.io/trusted-types/dist/spec/#htmlscriptelement-script-text</a> and <a href="https://w3c.github.io/trusted-types/dist/spec/#slot-value-verification">https://w3c.github.io/trusted-types/dist/spec/#slot-value-verification</a>

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-Node-multiple-arguments-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-text-node-insertion-into-script-element-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-report-only-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-svg-script-expected.txt:
* Source/WebCore/dom/ScriptElement.cpp:
(WebCore::ScriptElement::finishParsingChildren):
(WebCore::ScriptElement::prepareScript):
(WebCore::ScriptElement::setTrustedScriptText):
* Source/WebCore/dom/ScriptElement.h:
* Source/WebCore/html/HTMLScriptElement.cpp:
(WebCore::HTMLScriptElement::finishParsingChildren):
(WebCore::HTMLScriptElement::setTextContent):
(WebCore::HTMLScriptElement::setInnerText):
* Source/WebCore/html/HTMLScriptElement.h:
* Source/WebCore/svg/SVGScriptElement.cpp:
(WebCore::SVGScriptElement::finishParsingChildren):
* Source/WebCore/svg/SVGScriptElement.h:

Canonical link: <a href="https://commits.webkit.org/279117@main">https://commits.webkit.org/279117@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fcb3d806cce8d6f82cef6c294c69071951dbafd3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/52381 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/31713 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/4802 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/55655 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/3104 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/54686 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/38260 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/2803 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/42596 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/1979 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/54477 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/29421 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/45209 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/23677 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/26614 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/2484 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/1263 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/48507 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/2634 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/57251 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/27507 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/2665 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/49985 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/28740 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/45329 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/49235 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11479 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/29652 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/28485 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/dom/ScriptElement.h:
```diff
@@ -105,9 +105,13 @@ class ScriptElement {
     void handleAsyncAttribute();
 
+    // https://w3c.github.io/trusted-types/dist/spec/#slots-with-trusted-values
+    String m_scriptText { emptyString() };
```
Why is this public? It should be protected at worst.

lukewarlow replied:
Ah apologies that's an oversight on my part it should be protected as you say

darinadler replied:
Any reason to not let this be null string instead of empty string?

darinadler replied:
@rniwa, it is protected.

But since the derived class only needs to **set** `m_scriptText`, I think there should be a protected **function** for it to call, like maybe named something like `setTrustedScriptText` and `m_scriptText` itself can be private. I also think that `m_scriptText` should be named `m_trustedScriptText` for greater clarity.

lukewarlow replied:
> Any reason to not let this be null string instead of empty string?

Yes there is, I've just realised there isn't a web platform test that covers the difference so I will add one.

Effectively if it's a null string then an empty script created through DOM APIs would lead to a TT violation (because the script content is never a null string).

lukewarlow replied:
Turns out the existing tests do cover this just not super explicitly. It's also quite hard to test explicitly because you're testing for the lack of an event.

rniwa reviewed (approved):
r=me with the change to make the member variable protected.

darinadler commented on Source/WebCore/dom/ScriptElement.cpp:
```diff
@@ -170,7 +176,21 @@ bool ScriptElement::prepareScript(const TextPosition& scriptStartPosition)
-    auto sourceText = scriptContent();
+    String sourceText;
+
+    if (element().scriptExecutionContext()->settingsValues().trustedTypesEnabled) {
```
I think this would read nicer if we put `*element().scriptExecutionContext()` into a local variable.

darinadler commented on Source/WebCore/dom/ScriptElement.cpp:
```diff
@@ -170,7 +176,21 @@ bool ScriptElement::prepareScript(const TextPosition& scriptStartPosition)
         m_forceAsync = true;
 
-    auto sourceText = scriptContent();
+    String sourceText;
```
Both code paths below call scriptContent(), so I think we should structure the code like this:
```
String sourceText = scriptContent();
Ref context = *element().scriptExecutionContext();
if (context->settingsValues().trustedTypesEnabled && sourceText != m_scriptText) {
    auto trustedText = trustedTypeCompliantString(TrustedType::TrustedScript, context, childTextContent, is<HTMLScriptElement>(element()) ? "HTMLScriptElement text"_s : "SVGScriptElement text"_s);
    if (trustedText.hasException())
        return false;
    sourceText = trustedText.releaseReturnValue();
}
```

darinadler commented on Source/WebCore/html/HTMLScriptElement.cpp:
```diff
@@ -108,7 +114,10 @@ ExceptionOr<void> HTMLScriptElement::setTextContent(ExceptionOr<String> value)
+    auto newValue = value.releaseReturnValue();
+
+    m_scriptText = newValue;
+    setTextContent(WTFMove(newValue));
```
I’m a little surprised that we set `m_scriptText` _exactly_ the same way when calling `setTextContent` or `setInnerText`. Do those two functions do exactly the same thing? If so, then why doesn’t one call the other? If not, then why does `m_scriptText` get set exactly the same?

lukewarlow replied:
So I can answer the why. That's how it's specced and how Chromium has implemented it. As for whether they behave differently I'm not entirely sure, I know .text and .textContent are identical, .innerText has a more complicated spec but I'm not sure if those differences are meaningful wrt script elements. I'll make a spec issue to see if this is an oversight or not.

lukewarlow replied:
Raised issue https://github.com/w3c/trusted-types/issues/517

lukewarlow committed:
Implement trusted types enforcement for script elements
https://bugs.webkit.org/show_bug.cgi?id=269365

Reviewed by Darin Adler and Ryosuke Niwa.

Add script text string to script element. This is used to ensure that script elements can't be manipulated in untrusted ways.

See https://w3c.github.io/trusted-types/dist/spec/#htmlscriptelement-script-text and https://w3c.github.io/trusted-types/dist/spec/#slot-value-verification

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-Node-multiple-arguments-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-text-node-insertion-into-script-element-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-report-only-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-svg-script-expected.txt:
* Source/WebCore/dom/ScriptElement.cpp:
(WebCore::ScriptElement::finishParsingChildren):
(WebCore::ScriptElement::prepareScript):
(WebCore::ScriptElement::setTrustedScriptText):
* Source/WebCore/dom/ScriptElement.h:
* Source/WebCore/html/HTMLScriptElement.cpp:
(WebCore::HTMLScriptElement::finishParsingChildren):
(WebCore::HTMLScriptElement::setTextContent):
(WebCore::HTMLScriptElement::setInnerText):
* Source/WebCore/html/HTMLScriptElement.h:
* Source/WebCore/svg/SVGScriptElement.cpp:
(WebCore::SVGScriptElement::finishParsingChildren):
* Source/WebCore/svg/SVGScriptElement.h:

Canonical link: https://commits.webkit.org/279117@main
